mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,110 +19,446 @@
-            break
-        fi
-    done <<< "${commits}"
-
-    git checkout ${currBranch}
-
-    # we must reset Godeps.json to what it looked like in the kube-sync branch
-    # before the first commit that's to be cherry-picked so that any new
-    # Godep.json changes from k8s.io/kubernetes will apply cleanly. Note that 
-    # entries for k8s.io/* will be updated later in the process.
-    if [ "${cherrypicks_change_godeps}" = "true" ] && [ "${new_repo}" = "false" ]; then
-       local cleanGodepJsonCommit=${previousBranchSHA}
-       git checkout ${cleanGodepJsonCommit} Godeps/Godeps.json
-       if git diff --cached --exit-code &>/dev/null; then
-           echo "no need to reset Godeps.json!"
-       else
-           git -c user.name="Kubernetes Publisher" -c user.email="k8s-publish-robot@users.noreply.github.com" commit -m "sync: reset Godeps.json" -- Godeps/Godeps.json
-       fi
-    fi
-
-    echo "commits to be cherry-picked:"
-    echo "${commits}"
-    echo ""
-    
-    while read commitSHA; do
-        if [[ -z "${commitSHA}" ]]; then
-            continue
+        echo "Using branch point ${k_branch_point_commit} for new branch ${dst_branch}."
+        git branch -f filtered-branch-base ${k_branch_point_commit} >/dev/null
+        extra_filter_branches="${extra_filter_branches} filtered-branch-base"
+    elif [ "${new_repo}" = false ]; then
+        if [ -z "${k_base_commit}" ]; then
+            # preserve the value from ./kubernetes-sha if it is set. Otherwise, use the Kubernetes-commit lines in the commit message
+            k_base_commit="$(last-kube-commit ${dst_branch} || true)"
+        fi
+        if [ -z "${k_base_commit}" ]; then
+            echo "Couldn't find a Kubernetes-commit sha in any commit on ${dst_branch}."
+            return 1
         fi
-        echo "working ${commitSHA}"
-        git -c user.name="Kubernetes Publisher" -c user.email="k8s-publish-robot@users.noreply.github.com" cherry-pick --keep-redundant-commits ${commitSHA}
-    done <<< "${commits}"
-    
-    # track the k8s.io/kubernetes commit SHA so we can always determine which level of kube this repo matches
-    # track the filtered branch commit SHA so that we can determine which commits need to be picked
-    echo ${newKubeSHA} > kubernetes-sha
-    if git diff --exit-code &>/dev/null; then
-        echo "SHAs haven't changed!"
-        return
+        local k_base_merge=$(git-find-merge ${k_base_commit} upstream-kube/${src_branch})
+        if [ -z "${k_base_merge}" ]; then
+            echo "Didn't find merge commit of k8s.io/kubernetes commit ${k_base_commit}. Odd."
+            return 1
+        fi
+        git branch -f filtered-branch-base ${k_base_merge} >/dev/null
+        extra_filter_branches="${extra_filter_branches} filtered-branch-base"
+    fi
+
+    # this command rewrites git history to *only* include $subdirectory
+    echo "Rewriting upstream branch ${src_branch} to only include commits for ${subdirectory}."
```

Comment:
we could yes. Not sure it makes it so much simpler though. Maybe we could create `strategy` variable and use a simple case clause instead of repeated if clauses.

Reply:
The difficulty I experienced when reading the code is that I first read tens of lines of code for condition 1, then read tens of lines of code for condition 2, then I need to read another tens of line of code for condition 1, and I need to look back to recall what state each local vairables end at last time.

Reply:
Will try your suggestion. Let's see how it looks then.

Reply:
done